// office-ui-fabric-react/lib-amd/components/Button/SplitButton/SplitButton.styles.js
define(["require", "exports", "../../../Styling", "../../../Utilities"], function (require, exports, Styling_1, Utilities_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.getStyles = Utilities_1.memoizeFunction(function (theme, customStyles) {
        var splitButtonStyles = {
            splitButtonContainer: {
                position: 'relative',
                display: 'inline-block',
                border: '1px solid transparent',
                ':hover': {
                    border: '1px solid'
                },
                ':focus': {
                    outline: 'none!important',
                    border: '1px solid'
                }
            },
            splitButtonMenuButton: {
                padding: '6px',
                height: 'auto',
                boxSizing: 'border-box',
                border: '1px solid transparent',
                outline: 'transparent',
                userSelect: 'none',
                display: 'inline-block',
                textDecoration: 'none',
                textAlign: 'center',
                cursor: 'pointer',
                verticalAlign: 'top',
                width: '32px'
            },
            splitButtonMenuButtonDisabled: {
                ':hover': {
                    cursor: 'default'
                }
            },
            splitButtonFlexContainer: {
                display: 'flex',
                height: '100%',
                flexWrap: 'nowrap',
                justifyContent: 'center',
                alignItems: 'center'
            },
        };
        return Styling_1.mergeStyleSets(splitButtonStyles, customStyles);
    });
});
