// office-ui-fabric-react/lib-amd/components/Button/ActionButton/ActionButton.styles.js
define(["require", "exports", "../../../Styling", "../../../Utilities", "../BaseButton.styles"], function (require, exports, Styling_1, Utilities_1, BaseButton_styles_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var DEFAULT_BUTTON_HEIGHT = '40px';
    var DEFAULT_PADDING = '0 4px';
    exports.getStyles = Utilities_1.memoizeFunction(function (theme, customStyles, focusInset, focusColor) {
        var baseButtonStyles = BaseButton_styles_1.getStyles(theme);
        var actionButtonStyles = {
            root: {
                borderWidth: '0',
                padding: DEFAULT_PADDING,
                height: DEFAULT_BUTTON_HEIGHT,
                color: theme.palette.neutralPrimary,
                backgroundColor: 'transparent'
            },
            rootHovered: {
                color: theme.palette.themePrimary,
                icon: {
                    color: theme.palette.themePrimary
                }
            },
            iconHovered: {
                color: theme.palette.themePrimary
            },
            rootPressed: {
                color: theme.palette.black,
            },
            rootExpanded: {
                color: theme.palette.themePrimary
            },
            iconPressed: {
                color: theme.palette.themeDarker
            },
            rootDisabled: {
                color: theme.palette.neutralTertiary,
                backgroundColor: 'transparent'
            },
            rootChecked: {
                color: theme.palette.black,
            },
            iconChecked: {
                color: theme.palette.themeDarker
            },
            flexContainer: {
                justifyContent: 'flex-start'
            },
            icon: {
                color: theme.palette.themeDarkAlt
            },
            iconDisabled: {
                color: 'inherit'
            },
            menuIcon: {
                color: theme.palette.neutralSecondary
            }
        };
        return Styling_1.mergeStyleSets(baseButtonStyles, actionButtonStyles, customStyles);
    });
});
